compress/zstd.c

```diff
@@ -82,7 +82,7 @@ static ComprHandle *compr_zstd_open(short level)
 {
   struct ZstdComprData *cdata = zstd_cdata_new();
 
-  cdata->buf = mutt_mem_calloc(1, ZSTD_compressBound(1024 * 128));
+  cdata->buf = mutt_mem_calloc(ZSTD_compressBound(1024 * 128), 1);
```

Comment:
I feel this is a bit cheating: we *know* we need `ZSTD_compressBound(1024 * 128)` **char**s (hence the `1`), but we're saving it to a `void*` object.

Wouldn't it be better to turn `buf` into a `char *` and use `MUTT_MEM_CALLOC(ZSTD_compressBound(1024 * 128), char)` ?

Reply:
> I feel this is a bit cheating: we _know_ we need `ZSTD_compressBound(1024 * 128)` **char**s (hence the `1`), but we're saving it to a `void*` object.

`sizeof(void) == 1` is a lovely GNU extension.  :-)

I used that `1` as a special case to denote that it's `void`.  Since we don't have GNU extensions enabled, we can't use

```c
cdata->buf = CALLOC(ZSTD_compressBound(1024 * 128), void);
```

but that's basically what I meant.

> 
> Wouldn't it be better to turn `buf` into a `char *` and use `MUTT_MEM_CALLOC(ZSTD_compressBound(1024 * 128), char)` ?

zstd's API uses `void*`, so I'd prefer using their same type.

```c
$ grepc ZSTD_compressCCtx /usr/include/
/usr/include/zstd.h:ZSTDLIB_API size_t ZSTD_compressCCtx(ZSTD_CCtx* cctx,
                                     void* dst, size_t dstCapacity,
                               const void* src, size_t srcSize,
                                     int compressionLevel);
```

Reply:
> zstd's API uses `void*`, so I'd prefer using their same type.\

I don't buy this for two reasons:

1. by the same reasoning, everything we `malloc` we should keep in a `void *`, so we can pass it to `free`. I think the idea here is that zstd doesn't care what the actual pointer points to. It *could* have used `char *`.
2. `src` is also a `const void *`, but we're fine passing it a `const char *`

Reply:
Hmmm, I was thinking that the difference was that the source was ours, while dst was zstd's (since they are the ones who write to it).

However, after thinking about it a little bit more, compressed stuff is actually treated as characters (compressed files are still files, after all), so yeah, I'm convinced: let's use char.

Reply:
Done.